app/services/spaces/duplicate_detector_service.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+module Spaces
+  class DuplicateDetectorService < ApplicationService
+    def initialize(space)
+      @title = space.title
+      @address = space.address
+      @post_number = space.post_number
+      super()
+    end
+
+    def call
+      potential_duplicates = find_potential_duplicates
+      return potential_duplicates if potential_duplicates.present?
+
+      nil
+    end
+
+    def find_potential_duplicates
+      duplicates = []
+
+      # If any full matches, just return those first:
+      duplicates << full_match
+      return duplicates.flatten if duplicates[0].present?
+
+      # If any matches by full address, return those without checking further:
+      duplicates << by_full_address
+      return duplicates.flatten.compact if duplicates[0].present?
+
+      # Then return any matches by title and/or post
+      duplicates << by_post_and_title
+      duplicates.flatten.compact.uniq
+    end
+
+    def full_match
+      return nil unless @title.present? && @address.present? && @post_number.present?
+
+      Space.where(
+        "title LIKE ?", "%#{@title}%"
```

Comment:
Could do `"title ILIKE ?", "%#{@title}%"` instead to match without case sensitivity

Reply:
Sweet! Will fix